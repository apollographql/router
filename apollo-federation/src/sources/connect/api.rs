use apollo_compiler::collections::IndexMap;
use apollo_compiler::Schema;
use itertools::Itertools;
use thiserror::Error;
use url::Url;

use super::ConnectSpec;
use super::Connector;

#[derive(Debug)]
pub enum BaseUrl {
    Static(Url),
    Dynamic(String),
}

#[derive(Debug, Error)]
pub enum BaseUrlError {
    #[error("Invalid schema: {0}")]
    ParseError(String),

    #[error("Could not find connectors: {0}")]
    ConnectorError(String),
}

pub fn base_urls_for_connectors(
    sdl: &str,
    subgraph_name: &str,
) -> Result<IndexMap<String, BaseUrl>, BaseUrlError> {
    let schema = Schema::builder()
        .adopt_orphan_extensions()
        .parse(sdl, "schema.graphql")
        .build()
        .map_err(|e| BaseUrlError::ParseError(e.errors.iter().join("\n")))?;

    let Some((spec, _)) = ConnectSpec::get_from_schema(&schema) else {
        return Ok(Default::default());
    };

    let connectors = Connector::from_schema(&schema, subgraph_name, spec)
        .map_err(|e| BaseUrlError::ConnectorError(e.to_string()))?;

    let mut base_urls = IndexMap::default();

    for (_, connector) in connectors {
        if let (Some(base_url), Some(source_name)) =
            (connector.transport.source_url, connector.id.source_name)
        {
            // TODO: dynamic url strings when supported
            base_urls.insert(source_name, BaseUrl::Static(base_url));
        }
    }

    Ok(base_urls)
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    #[test]
    fn test_base_urls_for_connectors() {
        let sdl = r#"
        extend schema
          @link(url: "https://specs.apollo.dev/federation/v2.10")
          @link(
            url: "https://specs.apollo.dev/connect/v0.1"
            import: ["@connect", "@source"]
          )
          @source(
            name: "one"
            http: { baseURL: "https://jsonplaceholder.typicode.com/" }
          )
          @source(
            name: "two"
            http: { baseURL: "https://api.stripe.com/v1/" }
          )

        type Query {
          a: String @connect(source: "one", http: { GET: "/" }, selection: "$")
          b: String @connect(source: "one", http: { GET: "/" }, selection: "$")
          c: String @connect(source: "two", http: { GET: "/" }, selection: "$")
          d: String @connect(http: { GET: "https://localhost/" }, selection: "$")
        }
    "#;

        let base_urls = super::base_urls_for_connectors(sdl, "test").unwrap();

        assert_debug_snapshot!(base_urls, @r#"
        {
            "one": Static(
                Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "jsonplaceholder.typicode.com",
                        ),
                    ),
                    port: None,
                    path: "/",
                    query: None,
                    fragment: None,
                },
            ),
            "two": Static(
                Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.stripe.com",
                        ),
                    ),
                    port: None,
                    path: "/v1/",
                    query: None,
                    fragment: None,
                },
            ),
        }
        "#);
    }
}
