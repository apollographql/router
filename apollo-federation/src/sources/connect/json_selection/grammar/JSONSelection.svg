<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="239" height="81">
   <defs>
      <style type="text/css">
    @namespace "http://www.w3.org/2000/svg";
    .line                 {fill: none; stroke: #332200; stroke-width: 1;}
    .bold-line            {stroke: #140E00; shape-rendering: crispEdges; stroke-width: 2;}
    .thin-line            {stroke: #1F1400; shape-rendering: crispEdges}
    .filled               {fill: #332200; stroke: none;}
    text.terminal         {font-family: Verdana, Sans-serif;
                            font-size: 12px;
                            fill: #140E00;
                            font-weight: bold;
                          }
    text.nonterminal      {font-family: Verdana, Sans-serif;
                            font-size: 12px;
                            fill: #1A1100;
                            font-weight: normal;
                          }
    text.regexp           {font-family: Verdana, Sans-serif;
                            font-size: 12px;
                            fill: #1F1400;
                            font-weight: normal;
                          }
    rect, circle, polygon {fill: #332200; stroke: #332200;}
    rect.terminal         {fill: #FFC34D; stroke: #332200; stroke-width: 1;}
    rect.nonterminal      {fill: #FFDF9E; stroke: #332200; stroke-width: 1;}
    rect.text             {fill: none; stroke: none;}
    polygon.regexp        {fill: #FFECC7; stroke: #332200; stroke-width: 1;}
  </style>
   </defs>
   <polygon points="9 17 1 13 1 21"/>
   <polygon points="17 17 9 13 9 21"/>
   <a xmlns:xlink="http://www.w3.org/1999/xlink"
      xlink:href="#NakedSubSelection"
      xlink:title="NakedSubSelection">
      <rect x="51" y="3" width="140" height="32"/>
      <rect x="49" y="1" width="140" height="32" class="nonterminal"/>
      <text class="nonterminal" x="59" y="21">NakedSubSelection</text>
   </a>
   <a xmlns:xlink="http://www.w3.org/1999/xlink"
      xlink:href="#PathSelection"
      xlink:title="PathSelection">
      <rect x="51" y="47" width="106" height="32"/>
      <rect x="49" y="45" width="106" height="32" class="nonterminal"/>
      <text class="nonterminal" x="59" y="65">PathSelection</text>
   </a>
   <path class="line"
         d="m17 17 h2 m20 0 h10 m140 0 h10 m-180 0 h20 m160 0 h20 m-200 0 q10 0 10 10 m180 0 q0 -10 10 -10 m-190 10 v24 m180 0 v-24 m-180 24 q0 10 10 10 m160 0 q10 0 10 -10 m-170 10 h10 m106 0 h10 m0 0 h34 m23 -44 h-3"/>
   <polygon points="229 17 237 13 237 21"/>
   <polygon points="229 17 221 13 221 21"/>
</svg>
