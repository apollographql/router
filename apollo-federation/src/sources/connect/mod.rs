#![allow(unused_imports)]

mod json_selection;
mod url_path_template;

pub use json_selection::ApplyTo;
pub use json_selection::ApplyToError;
pub use json_selection::JSONSelection;
pub use json_selection::Key;
pub use json_selection::PathSelection;
pub use json_selection::SubSelection;
pub use url_path_template::URLPathTemplate;
