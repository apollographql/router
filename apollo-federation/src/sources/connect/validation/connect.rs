//! Parsing and validation of `@connect` directives

use apollo_compiler::Name;
use apollo_compiler::Node;
use apollo_compiler::ast::FieldDefinition;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::ObjectType;
use itertools::Itertools;

use self::entity::validate_entity_arg;
use self::selection::Selection;
use super::Code;
use super::Message;
use super::coordinates::ConnectDirectiveCoordinate;
use super::coordinates::connect_directive_name_coordinate;
use super::coordinates::source_name_value_coordinate;
use super::source::SourceName;
use crate::sources::connect::ConnectSpec;
use crate::sources::connect::id::ConnectedElement;
use crate::sources::connect::id::ObjectCategory;
use crate::sources::connect::spec::schema::CONNECT_SOURCE_ARGUMENT_NAME;
use crate::sources::connect::validation::connect::http::Http;
use crate::sources::connect::validation::graphql::SchemaInfo;

mod entity;
mod http;
mod selection;

pub(super) fn fields_seen_by_all_connects(
    schema: &SchemaInfo,
    all_source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    schema
        .types
        .values()
        .filter(|ty| !ty.is_built_in())
        .for_each(|extended_type| {
            if let ExtendedType::Object(node) = extended_type {
                match fields_seen_by_connectors_on_types(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }

                match fields_seen_by_connectors_on_fields(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }
            }
        });

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on types are valid
fn fields_seen_by_connectors_on_types(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let connect_directives = object
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // TODO: find a better place for feature gates like this
    if schema.connect_link.spec == ConnectSpec::V0_1 {
        return Err(vec![Message {
            code: Code::FeatureUnavailable,
            message: format!(
                "Using `@{connect_directive_name}` on `type {object_name}` requires connectors v0.2. Learn more at https://go.apollo.dev/connectors/changelog.",
                object_name = object.name,
                connect_directive_name = schema.connect_directive_name(),
            ),
            locations: object
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        }]);
    }

    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Type { type_def: object },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                messages.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => messages.push(error),
        }

        // TODO: validate batch/this arguments as key fields

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    messages.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                messages.extend(errs);
                continue;
            }
        };
        messages.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on object fields are valid
fn fields_seen_by_connectors_on_fields(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let object_category = if schema
        .schema_definition
        .query
        .as_ref()
        .is_some_and(|query| query.name == object.name)
    {
        ObjectCategory::Query
    } else if schema
        .schema_definition
        .mutation
        .as_ref()
        .is_some_and(|mutation| mutation.name == object.name)
    {
        ObjectCategory::Mutation
    } else {
        ObjectCategory::Other
    };
    let mut seen_fields = Vec::new();
    let mut messages = Vec::new();
    for field in object.fields.values() {
        match fields_seen_by_connector(field, object_category, source_names, object, schema) {
            Ok(fields) => seen_fields.extend(fields),
            Err(errs) => messages.extend(errs),
        }
    }
    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

fn fields_seen_by_connector(
    field: &Component<FieldDefinition>,
    category: ObjectCategory,
    source_names: &[SourceName],
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let source_map = &schema.sources;
    let mut errors = Vec::new();
    let connect_directives = field
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // mark the field with a @connect directive as seen
    let mut seen_fields = vec![(object.name.clone(), field.name.clone())];

    // direct recursion isn't allowed, like a connector on User.friends: [User]
    if &object.name == field.ty.inner_named_type() {
        errors.push(Message {
            code: Code::CircularReference,
            message: format!(
                "Direct circular reference detected in `{}.{}: {}`. For more information, see https://go.apollo.dev/connectors/limitations#circular-references",
                object.name,
                field.name,
                field.ty
            ),
            locations: field.line_column_range(source_map).into_iter().collect(),
        });
    }

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Field {
                parent_type: object,
                parent_category: category,
                field_def: field,
            },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                errors.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => errors.push(error),
        }

        errors
            .extend(validate_entity_arg(field, connect_directive, object, schema, category).err());

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    errors.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                errors.extend(errs);
                continue;
            }
        };
        errors.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if errors.is_empty() {
        Ok(seen_fields)
    } else {
        Err(errors)
    }
}

pub(super) fn validate_source_name<'schema>(
    directive: &Node<Directive>,
    coordinate: &ConnectDirectiveCoordinate,
    source_names: &'schema [SourceName],
    schema: &SchemaInfo,
) -> Result<Option<&'schema SourceName<'schema>>, Message> {
    let Some(source_name_arg) = directive
        .arguments
        .iter()
        .find(|arg| arg.name == CONNECT_SOURCE_ARGUMENT_NAME)
    else {
        return Ok(None);
    };

    let resolved_source_name = source_names
        .iter()
        .find(|name| **name == source_name_arg.value);

    if let Some(source_name) = resolved_source_name {
        return Ok(Some(source_name));
    }
    // A source name was set but doesn't match a defined source
    // TODO: Pick a suggestion that's not just the first defined source
    let qualified_directive = connect_directive_name_coordinate(
        schema.connect_directive_name(),
        &source_name_arg.value,
        coordinate,
    );
    if let Some(first_source_name) = source_names.first() {
        Err(Message {
            code: Code::SourceNameMismatch,
            message: format!(
                "{qualified_directive} does not match any defined sources. Did you mean \"{first_source_name}\"?",
                first_source_name = first_source_name.as_str(),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    } else {
        Err(Message {
            code: Code::NoSourcesDefined,
            message: format!(
                "{qualified_directive} specifies a source, but none are defined. Try adding {coordinate} to the schema.",
                coordinate = source_name_value_coordinate(
                    schema.source_directive_name(),
                    &source_name_arg.value
                ),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    }
}
