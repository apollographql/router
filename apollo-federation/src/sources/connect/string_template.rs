//! A [`StringTemplate`] is a string containing one or more [`Expression`]s.
//! These are used in connector URIs and headers.
//!
//! Parsing (this module) is done by both the router at startup and composition. Validation
//! (in [`crate::sources::connect::validation`]) is done only by composition.

use std::fmt::Display;
use std::ops::Range;
use std::str::FromStr;

use apollo_compiler::collections::IndexMap;
use itertools::Itertools;
use serde_json_bytes::Value;

use crate::sources::connect::JSONSelection;

/// A parsed string template, containing a series of [`Part`]s.
///
/// The `Const` generic allows consumers to validate constant pieces of the string with a type of
/// their choice. This is specifically just [`http::HeaderValue`] for headers right now.
#[derive(Clone, Debug)]
pub struct StringTemplate<Const = String> {
    pub(crate) parts: Vec<Part<Const>>,
}
impl<Const: FromStr> StringTemplate<Const> {
    /// Parse a [`StringTemplate`]. If this template is nested within another string, provide an
    /// `offset` to correct the locations.
    ///
    /// TODO: Remove the `offset` param once `URLTemplate` can leverage this more directly.
    pub(crate) fn parse(input: &str, mut offset: usize) -> Result<Self, Error> {
        let mut chars = input.chars().peekable();
        let mut parts = Vec::new();
        while let Some(next) = chars.peek() {
            if *next == '{' {
                let mut braces_count = 0; // Ignore braces within JSONSelection
                let expression = chars
                    .by_ref()
                    .skip(1)
                    .take_while(|c| {
                        if *c == '{' {
                            braces_count += 1;
                        } else if *c == '}' {
                            braces_count -= 1;
                        }
                        braces_count >= 0
                    })
                    .collect::<String>();
                if braces_count >= 0 {
                    return Err(Error {
                        message: "Invalid expression, missing closing }".into(),
                        location: offset..input.len(),
                    });
                }
                offset += 1; // Account for opening brace
                let parsed = JSONSelection::parse(&expression).map_err(|err| {
                    let start_of_parse_error = offset + err.offset;
                    Error {
                        message: err.message,
                        location: start_of_parse_error..(offset + expression.len()),
                    }
                })?;
                parts.push(Part::Expression(Expression {
                    expression: parsed,
                    location: offset..(offset + expression.len()),
                }));
                offset += expression.len() + 1; // Account for closing brace
            } else {
                let constant = chars
                    .by_ref()
                    .peeking_take_while(|c| *c != '{')
                    .collect::<String>();
                let value = Const::from_str(&constant).map_err(|_unhelpful_err| Error {
                    message: format!("invalid value `{constant}`"),
                    location: offset..offset + constant.len(),
                })?;
                parts.push(Part::Constant(Constant {
                    value,
                    location: offset..offset + constant.len(),
                }));
                offset += constant.len();
            }
        }
        Ok(Self { parts })
    }

    /// Get all the dynamic [`Expression`] pieces of the template for validation. If interpolating
    /// the entire template, use [`Self::interpolate`] instead.
    pub(crate) fn expressions(&self) -> impl Iterator<Item = &Expression> {
        self.parts.iter().filter_map(|part| {
            if let Part::Expression(expression) = part {
                Some(expression)
            } else {
                None
            }
        })
    }
}

impl StringTemplate<String> {
    /// Interpolation for when the constant type is a string. This can't be implemented for
    /// arbitrary generic types, so non-string consumers (headers) implement this themselves with
    /// any additional validations/transformations they need.
    pub(crate) fn interpolate(&self, vars: &IndexMap<String, Value>) -> Result<String, Error> {
        self.parts
            .iter()
            .map(|part| part.interpolate(vars))
            .collect()
    }
}

/// Expressions should be written the same as they were originally, even though we don't keep the
/// original source around. So constants are written as-is and expressions are surrounded with `{ }`.
impl<Const: Display> Display for StringTemplate<Const> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for part in &self.parts {
            match part {
                Part::Constant(Constant { value, .. }) => write!(f, "{}", value)?,
                Part::Expression(Expression { expression, .. }) => write!(f, "{{{}}}", expression)?,
            }
        }
        Ok(())
    }
}

/// A general-purpose error type which includes both a description of the problem and the offset span
/// within the original expression where the problem occurred. Used for both parsing and interpolation.
#[derive(Debug, PartialEq)]
pub struct Error {
    /// A human-readable description of the issue.
    pub message: String,
    /// The string offsets to the original [`StringTemplate`] (not just the part) where the issue
    /// occurred. As per usual, the end of the range is exclusive.
    pub(crate) location: Range<usize>,
}

impl Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for Error {}

/// One piece of a [`StringTemplate`]
#[derive(Clone, Debug)]
pub(crate) enum Part<Const> {
    /// A constant string literal—the piece of a [`StringTemplate`] _not_ in `{ }`
    Constant(Constant<Const>),
    /// A dynamic piece of a [`StringTemplate`], which came from inside `{ }` originally.
    Expression(Expression),
}

impl<T> Part<T> {
    /// Get the original location of the part from the string which was parsed to form the
    /// [`StringTemplate`].
    fn location(&self) -> Range<usize> {
        match self {
            Self::Constant(c) => c.location.clone(),
            Self::Expression(e) => e.location.clone(),
        }
    }
}

/// These generics are a bit of a mess, but what they're saying is given a generic `Const` type,
/// which again is `String` for the main use case but specialized occasionally (like [`http::HeaderValue`] for headers),
/// we can interpolate the value of the part into that type.
///
/// For [`Constant`]s this is easy, just clone the value (thus `Const: Clone`).
///
/// For [`Expression`]s, we first need to interpolate the expression as normal (with [`ApplyTo`]),
/// and then convert the resulting [`Value`] into the `Const` type. For that we require both
/// `Const: FromStr` and `Const: TryFrom<String>` so we don't have to clone all `&str` into `String`s,
/// nor borrow `String` just for them to be re-allocated. The `FromStrErr` and `TryFromStringError`
/// are then required to capture the error types of those two conversion methods.
///
/// So for `Const = String` these are actually all no-ops with infallible conversions, but we allow
/// for [`http::HeaderValue`] to fail.
impl<Const, FromStrErr, TryFromStringError> Part<Const>
where
    Const: Clone,
    Const: FromStr<Err = FromStrErr>,
    FromStrErr: std::error::Error,
    Const: TryFrom<String, Error = TryFromStringError>,
    TryFromStringError: std::error::Error,
{
    pub(crate) fn interpolate(&self, vars: &IndexMap<String, Value>) -> Result<Const, Error> {
        match self {
            Part::Constant(Constant { value, .. }) => Ok(value.clone()),
            Part::Expression(Expression { expression, .. }) => {
                // TODO: do something with the ApplyTo errors
                let (value, _errs) = expression.apply_with_vars(&Value::Null, vars);

                match value.unwrap_or(Value::Null) {
                    Value::Null => Const::from_str("").map_err(|err| Error {
                        message: err.to_string(),
                        location: self.location(),
                    }),
                    Value::Bool(b) => Const::try_from(b.to_string()).map_err(|err| Error {
                        message: err.to_string(),
                        location: self.location(),
                    }),
                    Value::Number(n) => Const::try_from(n.to_string()).map_err(|err| Error {
                        message: err.to_string(),
                        location: self.location(),
                    }),
                    Value::String(s) => Const::from_str(s.as_str()).map_err(|err| Error {
                        message: err.to_string(),
                        location: self.location(),
                    }),
                    Value::Array(_) | Value::Object(_) => Err(Error {
                        message: "Expressions can't evaluate to arrays or objects.".to_string(),
                        location: self.location(),
                    }),
                }
            }
        }
    }
}

/// A constant string literal—the piece of a [`StringTemplate`] _not_ in `{ }`
#[derive(Clone, Debug)]
pub(crate) struct Constant<T> {
    value: T,
    location: Range<usize>,
}

/// A dynamic piece of a [`StringTemplate`], which came from inside `{ }` originally.
#[derive(Clone, Debug)]
pub(crate) struct Expression {
    pub(crate) expression: JSONSelection,
    pub(crate) location: Range<usize>,
}

#[cfg(test)]
mod test_parse {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn simple_constant() {
        let template =
            StringTemplate::<String>::parse("text", 0).expect("simple template should be valid");
        assert_debug_snapshot!(template);
    }

    #[test]
    fn simple_expression() {
        assert_debug_snapshot!(StringTemplate::<String>::parse("{$config.one}", 0).unwrap());
    }
    #[test]
    fn mixed_constant_and_expression() {
        assert_debug_snapshot!(
            StringTemplate::<String>::parse("text{$config.one}text", 0).unwrap()
        );
    }

    #[test]
    fn offset() {
        assert_debug_snapshot!(
            StringTemplate::<String>::parse("text{$config.one}text", 9).unwrap()
        );
    }

    #[test]
    fn expressions_with_nested_braces() {
        assert_debug_snapshot!(
            StringTemplate::<String>::parse("const{$config.one { two { three } }}another-const", 0)
                .unwrap()
        );
    }

    #[test]
    fn missing_closing_braces() {
        assert_debug_snapshot!(
            StringTemplate::<String>::parse("{$config.one", 0),
            @r###"
        Err(
            Error {
                message: "Invalid expression, missing closing }",
                location: 0..12,
            },
        )
        "###
        )
    }
}

#[cfg(test)]
mod test_interpolate {
    use insta::assert_debug_snapshot;
    use pretty_assertions::assert_eq;
    use serde_json_bytes::json;

    use super::*;
    #[test]
    fn test_interpolate() {
        let template = StringTemplate::<String>::parse("before {$config.one} after", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": "foo"}));
        assert_eq!(template.interpolate(&vars).unwrap(), "before foo after");
    }

    #[test]
    fn test_interpolate_missing_value() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let vars = IndexMap::default();
        assert_eq!(template.interpolate(&vars).unwrap(), "");
    }

    #[test]
    fn test_interpolate_value_array() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": ["one", "two"]}));
        assert_debug_snapshot!(
            template.interpolate(&vars),
            @r###"
        Err(
            Error {
                message: "Expressions can't evaluate to arrays or objects.",
                location: 1..12,
            },
        )
        "###
        );
    }

    #[test]
    fn test_interpolate_value_bool() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": true}));
        assert_eq!(template.interpolate(&vars).unwrap(), "true");
    }

    #[test]
    fn test_interpolate_value_null() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": null}));
        assert_eq!(template.interpolate(&vars).unwrap(), "");
    }

    #[test]
    fn test_interpolate_value_number() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": 1}));
        assert_eq!(template.interpolate(&vars).unwrap(), "1");
    }

    #[test]
    fn test_interpolate_value_object() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": {}}));
        assert_debug_snapshot!(
            template.interpolate(&vars),
            @r###"
        Err(
            Error {
                message: "Expressions can't evaluate to arrays or objects.",
                location: 1..12,
            },
        )
        "###
        );
    }

    #[test]
    fn test_interpolate_value_string() {
        let template = StringTemplate::<String>::parse("{$config.one}", 0).unwrap();
        let mut vars = IndexMap::default();
        vars.insert("$config".to_string(), json!({"one": "string"}));
        assert_eq!(template.interpolate(&vars).unwrap(), "string");
    }
}

#[cfg(test)]
mod test_get_expressions {
    use super::*;

    #[test]
    fn test_variable_references() {
        let value =
            StringTemplate::<String>::parse("a {$this.a.b.c} b {$args.a.b.c} c {$config.a.b.c}", 0)
                .unwrap();
        let references: Vec<_> = value
            .expressions()
            .map(|e| e.expression.to_string())
            .collect();
        assert_eq!(
            references,
            vec!["$this.a.b.c", "$args.a.b.c", "$config.a.b.c"]
        );
    }
}
