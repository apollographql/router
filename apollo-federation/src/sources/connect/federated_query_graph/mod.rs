use apollo_compiler::NodeStr;
use indexmap::IndexMap;
use indexmap::IndexSet;

use crate::schema::position::EnumTypeDefinitionPosition;
use crate::schema::position::ObjectTypeDefinitionPosition;
use crate::schema::position::ScalarTypeDefinitionPosition;
use crate::schema::ObjectFieldDefinitionPosition;
use crate::sources::connect::json_selection::JSONSelection;
use crate::sources::connect::json_selection::Key;
use crate::sources::connect::json_selection::PathSelection;
use crate::sources::connect::json_selection::SubSelection;
use crate::sources::source;
use crate::sources::source::federated_query_graph::FederatedQueryGraphApi;
use crate::sources::source::SourceId;
use crate::ValidFederationSubgraph;

pub(crate) mod builder;

#[derive(Debug)]
pub(crate) struct FederatedQueryGraph {
    subgraphs_by_name: IndexMap<NodeStr, ValidFederationSubgraph>,
    // source_directives_by_name: IndexMap<NodeStr, SourceDirectiveArguments>,
    // connect_directives_by_source: IndexMap<ConnectId, ConnectDirectiveArguments>,
}

impl FederatedQueryGraphApi for FederatedQueryGraph {
    fn execution_metadata(
        &self,
    ) -> IndexMap<SourceId, source::query_plan::query_planner::ExecutionMetadata> {
        todo!()
    }
}

#[derive(Debug)]
pub(crate) struct AbstractNode;

#[derive(Debug)]
pub(crate) enum ConcreteNode {
    ConnectParent {
        subgraph_type: ObjectTypeDefinitionPosition,
    },
    SelectionRoot {
        subgraph_type: ObjectTypeDefinitionPosition,
        property_path: Vec<Key>,
    },
    SelectionChild {
        subgraph_type: ObjectTypeDefinitionPosition,
    },
}

#[derive(Debug)]
pub(crate) enum EnumNode {
    SelectionRoot {
        subgraph_type: EnumTypeDefinitionPosition,
        property_path: Vec<Key>,
    },
    SelectionChild {
        subgraph_type: EnumTypeDefinitionPosition,
    },
}

#[derive(Debug)]
pub(crate) enum ScalarNode {
    SelectionRoot {
        subgraph_type: ScalarTypeDefinitionPosition,
        property_path: Vec<Key>,
    },
    CustomScalarSelectionRoot {
        subgraph_type: ScalarTypeDefinitionPosition,
        selection: JSONSelection,
    },
    SelectionChild {
        subgraph_type: ScalarTypeDefinitionPosition,
    },
}

#[derive(Debug)]
pub(crate) struct AbstractFieldEdge;

#[derive(Debug)]

pub(crate) enum ConcreteFieldEdge {
    Connect {
        subgraph_field: ObjectFieldDefinitionPosition,
    },
    Selection {
        subgraph_field: ObjectFieldDefinitionPosition,
        property_path: Vec<Key>,
    },
    CustomScalarPathSelection {
        subgraph_field: ObjectFieldDefinitionPosition,
        path_selection: PathSelection,
    },
    CustomScalarStarSelection {
        subgraph_field: ObjectFieldDefinitionPosition,
        star_subselection: Option<SubSelection>,
        excluded_properties: IndexSet<Key>,
    },
}

#[derive(Debug)]
pub(crate) struct TypeConditionEdge;

#[derive(Debug)]
pub(crate) enum SourceEnteringEdge {
    ConnectParent {
        subgraph_type: ObjectTypeDefinitionPosition,
    },
}
