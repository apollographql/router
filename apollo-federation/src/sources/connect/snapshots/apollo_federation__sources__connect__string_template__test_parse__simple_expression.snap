---
source: apollo-federation/src/sources/connect/string_template.rs
expression: "StringTemplate::<String>::parse(\"{$config.one}\", 0).unwrap()"
---
StringTemplate {
    parts: [
        Expression(
            Expression {
                expression: Path(
                    PathSelection {
                        path: WithRange {
                            node: Var(
                                WithRange {
                                    node: $config,
                                    range: Some(
                                        0..7,
                                    ),
                                },
                                WithRange {
                                    node: Key(
                                        WithRange {
                                            node: Field(
                                                "one",
                                            ),
                                            range: Some(
                                                8..11,
                                            ),
                                        },
                                        WithRange {
                                            node: Empty,
                                            range: Some(
                                                11..11,
                                            ),
                                        },
                                    ),
                                    range: Some(
                                        7..11,
                                    ),
                                },
                            ),
                            range: Some(
                                0..11,
                            ),
                        },
                    },
                ),
                location: 1..12,
            },
        ),
    ],
}
